NIFI-9390: Updates to MergeContent / MergeRecord so that they play ni…

markap14 opened the pull request:
…cely within Stateless

<!--
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

turcsanyip commented:
@markap14 Thanks for implementing the changes in the processors. The merge processors now work properly both in traditional and stateless NiFi with simple flows.

However, there is still an issue in the stateless framework related to the execution order of the processors which becomes apparent when you have multiple (at least 3) processors before the merge processor. In this case, the expected order of execution would be:
- Proc1
- Proc2
- Proc3
- Merge (more input needed)
- Proc1
- Proc2
- Proc3
- Merge (more input needed)
- Proc1
- ...

But the actual order is:
- Proc1
- Proc2
- Proc3
- Merge (more input needed)
- Proc1
- Proc2
- Merge (no new input as Proc3 has not called yet, outputs the single item received in the 1st round)
- Proc3 (called in the wrong order)

The root cause seems to be that although `AsynchronousCommitTracker` registers the ready components in the right order: https://github.com/apache/nifi/blob/56593ad12f40bbacd4627436eadc38a24f58ee16/nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/session/AsynchronousCommitTracker.java#L47-L57 `StandardStatelessFlowCurrent` iterates over an earlier (and potentially outdated) snapshot of the ready components: https://github.com/apache/nifi/blob/56593ad12f40bbacd4627436eadc38a24f58ee16/nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/flow/StandardStatelessFlowCurrent.java#L72-L92 (The set and/or order of the ready components may be changed in `triggerWhileReady()` within the for loop)

My suggested solution would be to implement `Connectable getNextReady()` in `AsynchronousCommitTracker` instead of `List<Connectable> getReady()` and to call it from `StandardStatelessFlowCurrent` (and to remove the for loop).

What is your opinion about it?

turcsanyip commented:
@markap14 Yesterday I forgot to test the most trivial flow: Input Port => MergeContent (=> LogAttribute)
It does not work with this change because the root component (MergeContent in this case) does not get triggered any more when no input available but MergeContent would need an extra call with no FF now.
Could you please check this use case too?

markap14 commented:
Great catch @turcsanyip! I was able to add system tests that replicate the behavior and address the issues, I believe. I also cleaned up a bit of code and added javadocs to help clarify how the StandardStatelessFlowCurrent and associated ProgressTracker work together.

turcsanyip commented on nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/flow/StandardStatelessFlowCurrent.java:
```diff
@@ -57,23 +58,22 @@ private StandardStatelessFlowCurrent(final Builder builder) {
         try {
             boolean completionReached = false;
             while (!completionReached) {
+                tracker.shelveReadyComponents();
```
Is shelving the previously ready components really needed?

When I found the issue, I quickly fixed it with using the existing `ready` `LinkedHashSet` and implementing a simple `getNextReady()` method in `AsynchronousCommitTracker` which returns the top element of the ready "stack". I tested it again with several flows and could not find any issues with that. The advantage would be that the code is simpler and more straightforward (less datastructures, less loops).

At this point in the execution, the ready stack contains the component that triggered the root components for more input (at the top of the stack), and may contain some other already ready components. During the re-execution of the flow from the root components, these components may become ready again and they will be moved to the top of the stack or may just be waiting in the stack in their original position (as they were shelved). In both cases they will be triggered when their turn comes and will process all the FlowFiles from their input queue. That's why I think that shelving is not needed.

The code would look like this:
```
            while (!completionReached) {
                triggerRootConnectables();

                while (tracker.isAnyReady()) {
                    final Connectable connectable = tracker.getNextReady();
                    logger.debug("The next ready component to be triggered: {}", connectable);

                    // Continually trigger the given component as long as it is ready to be triggered
                    NextConnectable nextConnectable = triggerWhileReady(connectable);

                    // If there's nothing left to do, return
                    if (nextConnectable == NextConnectable.NONE) {
                        return;
                    }

                    // If next connectable is whatever is ready, just continue loop
                    if (nextConnectable == NextConnectable.NEXT_READY) {
                        continue;
                    }

                    // Otherwise, we need to break out of this loop so that we can trigger root connectables or complete dataflow
                    break;
                }

                // We have reached completion if the tracker does not know of any components ready to be triggered AND
                // we have no data queued in the flow (with the exception of Output Ports).
                completionReached = !tracker.isAnyReady() && isFlowQueueEmpty();
            }
```

markap14 replied:
@turcsanyip can you share the proposed code for a getNextReady() method for the `AsynchronousCommitTracker`?

turcsanyip replied:
It returns the last element from the ready list (unfortunately there is no "getLastElement()" in `LinkedHashSet`):
```
    public Connectable getNextReady() {
        if (ready.isEmpty()) {
            return null;
        }
        return new ArrayList<>(ready).get(ready.size() - 1);
    }
```

markap14 replied:
I see. The key here being that in this approach the getNextReady() would return the last element in the list, rather than the List of all ready components. I was missing that distinction at first. I think you're right - this should work essentially the same as the solution that I put in this PR, but your method is cleaner. Simpler. And possibly more efficient, as well.
Will update the PR to go this route. Thanks! I'll probably avoid creating the new ArrayList and returning the last element, though, and instead just iterate over the Set to get the last element - just to avoid that intermediate ArrayList. But otherwise I think your solution is really good. Nice job!

turcsanyip reviewed (changes requested):
@markap14 I tested the flow execution with different flows and cannot find any issues.
Thanks for the additional javadoc comments, mostly on the name of `StatelessFlowCurrent`!

As the behaviour of the Merge* processors has changed, the recommended configuration has changed too. Could you please update the sections about merging in the Stateless and Kafka Connector READMEs?

I added two other comments inline. One is about logging refinement, the other is a possible code logic simplification.
Could you please check them too?

turcsanyip reviewed (changes requested):
@markap14 I tested the flow execution with different flows and cannot find any issues.
Thanks for the additional javadoc comments, mostly on the name of `StatelessFlowCurrent`!

As the behaviour of the Merge* processors has changed, the recommended configuration has changed too. Could you please update the sections about merging in the Stateless and Kafka Connector READMEs?

I added two other comments inline. One is about logging refinement, the other is a possible code logic simplification.
Could you please check them too?

markap14 commented on nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/session/AsynchronousCommitTracker.java:
```diff
@@ -89,7 +117,7 @@ public boolean isReady(final Connectable connectable) {
         }
 
-        if (isDataQueued(connectable)) {
+        if (isDataQueued(connectable) || (connectable.isTriggerWhenEmpty() && isDataHeld(connectable))) {
             logger.debug("{} {} is ready because it has data queued", this, connectable);
             return true;
         }
```
Yeah good call. If we run into issues that may be helpful. Will update that.

markap14 commented:
OK @turcsanyip I pushed a new commit to incorporate the new logic. All the system tests, etc. pass. I also rebased against main so had to do a force-push.

turcsanyip reviewed (approved):
@markap14 Thanks for the changes, everything looks good now.
Merging to main.

markap14 commented:
Thanks @turcsanyip !
